horovod/mxnet/mpi_ops.cc

```diff
@@ -49,206 +51,154 @@ inline void InvokeCompleteCallback(CallbackOnComplete on_complete, const Status&
   }
 }
 
-void DoAllreduce(NDArray* tensor, NDArray* output, const std::string& name,
-                 CallbackOnComplete on_complete) {
-  ThrowIfError(common::CheckInitialized());
-
-  auto device = TensorUtil::GetDevice(tensor);
-  auto hvd_tensor = std::make_shared<MXTensor<NDArray>>(tensor);
-  auto hvd_context = std::make_shared<MXOpContext<NDArray>>(device, output);
-  auto hvd_output = std::make_shared<MXTensor<NDArray>>(output);
-
-  auto enqueue_result = EnqueueTensorAllreduce(
-      hvd_context, hvd_tensor, hvd_output, nullptr, name, device,
-      [on_complete](const Status& status) {
-        InvokeCompleteCallback(on_complete, status);
-      });
-  ThrowIfError(enqueue_result);
+inline const char* GetOpTypeName(OperationType op_type) {
+  switch (op_type) {
+    case OperationType::ALLREDUCE:
+      return ALLREDUCE_OP_TYPE_NAME;
+    case OperationType::ALLGATHER:
+      return ALLGATHER_OP_TYPE_NAME;
+    case OperationType::BROADCAST:
+      return BROADCAST_OP_TYPE_NAME;
+    default:
+      throw std::logic_error("Unsupported Horovod operation type.");
+  }
 }
 
-void DoAllgather(NDArray* tensor, NDArray* output, std::string& name,
-                 CallbackOnComplete on_complete) {
+void DoHorovodOperation(void*, void* on_complete_ptr, void* param) {
```

Comment:
Do we still need the first argument in the function?

Reply:
Answered below.